src/transports/ipc.rs

```diff
@@ -218,31 +246,65 @@ async fn run_server(unix_stream: UnixStream, messages_rx: mpsc::UnboundedReceive
     }
 }
 
+fn notify(
+    subscription_txs: &mut BTreeMap<SubscriptionId, mpsc::UnboundedSender<rpc::Value>>,
+    notification: rpc::Notification,
+) -> std::result::Result<(), ()> {
+    if let rpc::Params::Map(params) = notification.params {
+        let id = params.get("subscription");
+        let result = params.get("result");
+
+        if let (Some(&rpc::Value::String(ref id)), Some(result)) = (id, result) {
+            let id: SubscriptionId = id.clone().into();
+            if let Some(tx) = subscription_txs.get(&id) {
+                if let Err(e) = tx.send(result.clone()) {
+                    log::error!("Error sending notification: {:?} (id: {:?}", e, id);
+                }
+            } else {
+                log::warn!("Got notification for unknown subscription (id: {:?})", id);
+            }
+        } else {
+            log::error!("Got unsupported notification (id: {:?})", id);
+        }
+    }
+
+    Ok(())
+}
+
 fn respond(
     pending_response_txs: &mut BTreeMap<RequestId, oneshot::Sender<rpc::Value>>,
-    output: rpc::Output,
+    response: rpc::Response,
 ) -> std::result::Result<(), ()> {
-    let id = output.id().clone();
+    let outputs = match response {
+        rpc::Response::Single(output) => vec![output],
+        rpc::Response::Batch(outputs) => outputs,
+    };
 
-    let value = helpers::to_result_from_output(output).map_err(|err| {
-        log::warn!("Unable to parse output into rpc::Value: {:?}", err);
-    })?;
+    for output in outputs {
```

Comment:
This changes semantics of batch responses quite significantly actually.
Previously in case of error with single `output` you'd continue with others - now we will early exit the function.
Please consider extracting the inner-part of the loop to a scoped function or something, so that we conitnue with other outputs in case of a failure.
(might be good to add a test for it too :))

Reply:
You are right! We didn't realise this at the time, but definitely going to fix (and test) it. Brb

Reply:
@tomusdrw Just amended the fix and added a test that would break had the fix not been made!